vertical-pod-autoscaler/pkg/recommender/routines/replica_restrictions_post_processor.go

```diff
@@ -0,0 +1,208 @@
+// In other words:
+// - If we have less than <lowWatermark> replicas, we will allow the pods to be smaller, the intent it to keep the number
+//   of pods above <lowWatermark>. This allows you to improve resource utilisation to avoid having few very big and underutilised pods.
+// - If we have more than <highWatermark> replicas, we will allow the pods to be bigger, the intent it to keep the number
+//   of pods below <highWatermark>. This allows you to reduce the pressure on the k8s control plane.
+// If the annotation is defined and set to an empty value, the default value is used: [10, 100].
+// Any increase of resources is considered an upscale, any decrease of resources is considered a downscale.
+// This is best used with the `maintainedRatio` policy or with a single controlled resource.
+type ReplicaRestrictionsPostProcessor struct {
+	ControllerFetcher controllerfetcher.ControllerFetcher
+}
+
+// Process applies the Resource Ratio post-processing to the recommendation.
+func (r *ReplicaRestrictionsPostProcessor) Process(vpa *model.Vpa, recommendation *vpa_types.RecommendedPodResources, _ *vpa_types.PodResourcePolicy) *vpa_types.RecommendedPodResources {
+	watermarks := readWatermarksFromVPAAnnotations(vpa)
+	if watermarks == nil || len(watermarks) == 0 || recommendation == nil {
+		return recommendation
+	}
+	lowWatermark, highWatermark := watermarks[0], watermarks[1]
+
+	podTemplate, err := r.ControllerFetcher.GetPodTemplateFromTopMostWellKnown(&controllerfetcher.ControllerKeyWithAPIVersion{
+		ControllerKey: controllerfetcher.ControllerKey{
+			Namespace: vpa.ID.Namespace,
+			Kind:      vpa.TargetRef.Kind,
+			Name:      vpa.TargetRef.Name,
+		},
+		ApiVersion: vpa.TargetRef.APIVersion,
+	})
+	if err != nil {
+		klog.Errorf("Failed to apply ReplicaRestrictionsPostProcessor (controller fetch) to vpa %s/%s due to error: %#v", vpa.ID.Namespace, vpa.ID.VpaName, err)
+		return recommendation
+	}
+
+	pod := newPodFromTemplate(podTemplate)
+
+	updatedRecommendation, err := r.apply(recommendation, vpa.PodCount, lowWatermark, highWatermark, pod)
+	if err != nil {
+		klog.Errorf("Failed to apply ReplicaRestrictionsPostProcessor to vpa %s/%s due to error: %#v", vpa.ID.Namespace, vpa.ID.VpaName, err)
+	}
+	return updatedRecommendation
+}
+
+func readWatermarksFromVPAAnnotations(vpa *model.Vpa) []int {
+	var watermarks []int
+
+	for key, value := range vpa.Annotations {
+		if key != vpaPostProcessorPrefix + vpaPostProcessorReplicaRestrictionsSuffix {
+			continue
+		}
+
+		if value == "" {
+			watermarks = []int{10, 100}
+		} else if err := json.Unmarshal([]byte(value), &watermarks); err != nil {
+			klog.Errorf("Skipping restrictions definition '%s' in vpa %s/%s due to bad format, error:%#v", value, vpa.ID.Namespace, vpa.ID.VpaName, err)
+		}
+		if len(watermarks) != 2 {
+			klog.Errorf("Skipping restrictions definition '%s' in vpa %s/%s due to bad format", value, vpa.ID.Namespace, vpa.ID.VpaName)
+			continue
+		}
```

Comment:
In both these cases, we should probably reset `watermarks` to `int{}` and return early. For example, if a user gives an input with one or three numbers, i think we'll trigger the error, but we'll still return a none empty array.

Reply:
changed